sklearn/linear_model/_glm/glm.py

```diff
@@ -0,0 +1,622 @@
+                         tol=tol, warm_start=warm_start, verbose=verbose)
+
+    @property
+    def family(self):
+        # We use a property with a setter, since the GLM solver relies
+        # on self.family attribute, but we can't set it in __init__ according
+        # to scikit-learn API constraints. This attribute is made read-only
+        # to disallow changing distribution to other than Poisson.
+        return "poisson"
+
+    @family.setter
+    def family(self, value):
+        if value != "poisson":
+            raise ValueError("PoissonRegressor.family must be 'poisson'!")
+
+
+class GammaRegressor(GeneralizedLinearRegressor):
+    """Generalized Linear Model with a Gamma distribution.
+
+    Read more in the :ref:`User Guide <Generalized_linear_regression>`.
+
+    Parameters
+    ----------
+    alpha : float, default=1
+        Constant that multiplies the penalty term and thus determines the
+        regularization strength. ``alpha = 0`` is equivalent to unpenalized
+        GLMs. In this case, the design matrix `X` must have full column rank
+        (no collinearities).
+
+    fit_intercept : bool, default=True
+        Specifies if a constant (a.k.a. bias or intercept) should be
+        added to the linear predictor (X*coef+intercept).
+
+    max_iter : int, default=100
+        The maximal number of iterations for the solver.
+
+    tol : float, default=1e-4
+        Stopping criterion. For the lbfgs solver,
+        the iteration will stop when ``max{|g_i|, i = 1, ..., n} <= tol``
+        where ``g_i`` is the i-th component of the gradient (derivative) of
+        the objective function.
+
+    warm_start : bool, default=False
+        If set to ``True``, reuse the solution of the previous call to ``fit``
+        as initialization for ``coef_`` and ``intercept_`` .
+
+    verbose : int, default=0
+        For the lbfgs solver set verbose to any positive number for verbosity.
+
+    Attributes
+    ----------
+    coef_ : array of shape (n_features,)
+        Estimated coefficients for the linear predictor (X*coef_+intercept_) in
+        the GLM.
+
+    intercept_ : float
+        Intercept (a.k.a. bias) added to linear predictor.
+
+    n_iter_ : int
+        Actual number of iterations used in the solver.
+    """
+    def __init__(self, *, alpha=1.0, fit_intercept=True, max_iter=100,
+                 tol=1e-4, warm_start=False, verbose=0):
+
+        super().__init__(alpha=alpha, fit_intercept=fit_intercept,
+                         family="gamma", link='log', max_iter=max_iter,
+                         tol=tol, warm_start=warm_start, verbose=verbose)
+
+    @property
+    def family(self):
+        # We use a property with a setter, since the GLM solver relies
+        # on self.family attribute, but we can't set it in __init__ according
+        # to scikit-learn API constraints. This attribute is made read-only
+        # to disallow changing distribution to other than Gamma.
```

Comment:
> In particular I don't understand the part that says " we can't set it in __init__": the family attribute is set in __init__, since we call super.

Basically we needs `estimator.family` to exist at fit, because it's used in the parent class (and defines the behavior of fit). However if we do something like,
```py
class CustomRegressor(DummyRegressor): 
     def __init__(self): 
         super().__init__() 
         self.family = 'poisson'
```
this would fail `check_no_attributes_set_in_init` of `check_estimator` with,
```
AssertionError: Estimator CustomRegressor should not set any attribute apart from parameters during init. Found attributes ['family'].
```
So using a property with a setter is a way to make `check_estimator` happy. It `self.family = ..` is indeed called in super(), but it never creates that attribute and instead calls the setter. If you say it's a hack around scikit-learn API constraints, I agree :)

Reply:
@rth I think `check_estimator` would pass even without the use of a `property` here, since the common check allows for the attributes to be set in the parent class' `__init__`, as long as they are `__init__` parameters of the parent class:

```py
    invalid_attr = (set(vars(estimator)) - set(init_params)
                    - set(parents_init_params))
```

In fact, if you comment out the property and the setter for `GammaRegressor`, the common checks will run just fine (obligatory "at least it works on my machine"). Of course the `test_gamma_regression_family` about making the attribute read-only will fail though, but this one is not part of the common checks.


Basically, I think the only reason we need the property is to make the attributes read-only. And it's a good reason, I'm not saying we should remove the property. I'm just trying to understand what I'm missing but if I'm not, maybe we can simplify the comment to just say "We need the attribute to be read-only to avoid mis-uses e.g. in GridSearch."

Reply:
If it's an init property of the parent, why aren't we passing it in the super.init call? If it's not, can we use a class attribute rather than a property?

Reply:
> check_estimator would pass even without the use of a property here, since the common check allows for the attributes to be set in the parent class' __init__, as long as they are __init__ parameters of the parent class:

Ahh, that is great, thanks for investigating. I agree with you, it still wouldn't hurt to have read-only attributes, will update the comment as suggested.

> If it's an init property of the parent, why aren't we passing it in the super.init call?

We are passing it to `super().__init__` (see diff above) my previous example was just overly simplified to show that.